File: modules/textanalyzer-remote/src/main/java/org/opencastproject/textanalyzer/remote/TextAnalysisRemoteImpl.java
/**
 * Licensed to The Apereo Foundation under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 *
 * The Apereo Foundation licenses this file to you under the Educational
 * Community License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License
 * at:
 *
 *   http://opensource.org/licenses/ecl2.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 *
 */

package org.opencastproject.textanalyzer.remote;

import org.opencastproject.job.api.Job;
import org.opencastproject.job.api.JobParser;
import org.opencastproject.mediapackage.Attachment;
import org.opencastproject.mediapackage.MediaPackageElementParser;
import org.opencastproject.security.api.TrustedHttpClient;
import org.opencastproject.serviceregistry.api.RemoteBase;
import org.opencastproject.serviceregistry.api.ServiceRegistry;
import org.opencastproject.textanalyzer.api.TextAnalyzerException;
import org.opencastproject.textanalyzer.api.TextAnalyzerService;

import org.apache.http.HttpResponse;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.message.BasicNameValuePair;
import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Reference;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

@Component(
    immediate = true,
    service = TextAnalyzerService.class,
    property = {
        "service.description=Text Analysis Remote Service Proxy"
    }
)
public class TextAnalysisRemoteImpl extends RemoteBase implements TextAnalyzerService {

  /** The logger */
  private static final Logger logger = LoggerFactory.getLogger(TextAnalysisRemoteImpl.class);

  public TextAnalysisRemoteImpl() {
    super(JOB_TYPE);
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public Job extract(final Attachment image) throws TextAnalyzerException {
    HttpPost post = new HttpPost();
    try {
      List<BasicNameValuePair> params = new ArrayList<BasicNameValuePair>();
      params.add(new BasicNameValuePair("image", MediaPackageElementParser.getAsXml(image)));
      post.setEntity(new UrlEncodedFormEntity(params, "UTF-8"));
    } catch (Exception e) {
      throw new TextAnalyzerException(e);
    }
    HttpResponse response = null;
    try {
      response = getResponse(post);
      if (response != null) {
        try {
          Job receipt = JobParser.parseJob(response.getEntity().getContent());
          logger.info("Analyzing {} on a remote analysis server", image);
          return receipt;
        } catch (Exception e) {
          throw new TextAnalyzerException("Unable to analyze element '" + image + "' using a remote analysis service",
                  e);
        }
      }
    } finally {
      closeConnection(response);
    }
    throw new TextAnalyzerException("Unable to analyze element '" + image + "' using a remote analysis service");
  }

  @Reference(name = "trustedHttpClient")
  @Override
  public void setTrustedHttpClient(TrustedHttpClient trustedHttpClient) {
    super.setTrustedHttpClient(trustedHttpClient);
  }

  @Reference(name = "remoteServiceManager")
  @Override
  public void setRemoteServiceManager(ServiceRegistry serviceRegistry) {
    super.setRemoteServiceManager(serviceRegistry);
  }

}

File: modules/textanalyzer-remote/src/main/java/org/opencastproject/textanalyzer/remote/TextAnalysisRemoteImpl.kt
/**
 * Licensed to The Apereo Foundation under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 *
 * The Apereo Foundation licenses this file to you under the Educational
 * Community License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License
 * at:
 *
 * http://opensource.org/licenses/ecl2.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 *
 */

package org.opencastproject.textanalyzer.remote

import org.opencastproject.job.api.Job
import org.opencastproject.job.api.JobParser
import org.opencastproject.mediapackage.Attachment
import org.opencastproject.mediapackage.MediaPackageElementParser
import org.opencastproject.serviceregistry.api.RemoteBase
import org.opencastproject.textanalyzer.api.TextAnalyzerException
import org.opencastproject.textanalyzer.api.TextAnalyzerService

import org.apache.http.HttpResponse
import org.apache.http.client.entity.UrlEncodedFormEntity
import org.apache.http.client.methods.HttpPost
import org.apache.http.message.BasicNameValuePair
import org.slf4j.Logger
import org.slf4j.LoggerFactory

import java.util.ArrayList

class TextAnalysisRemoteImpl : RemoteBase(JOB_TYPE), TextAnalyzerService {

    /**
     * {@inheritDoc}
     */
    @Throws(TextAnalyzerException::class)
    override fun extract(image: Attachment): Job {
        val post = HttpPost()
        try {
            val params = ArrayList<BasicNameValuePair>()
            params.add(BasicNameValuePair("image", MediaPackageElementParser.getAsXml(image)))
            post.entity = UrlEncodedFormEntity(params, "UTF-8")
        } catch (e: Exception) {
            throw TextAnalyzerException(e)
        }

        var response: HttpResponse? = null
        try {
            response = getResponse(post)
            if (response != null) {
                try {
                    val receipt = JobParser.parseJob(response.entity.content)
                    logger.info("Analyzing {} on a remote analysis server", image)
                    return receipt
                } catch (e: Exception) {
                    throw TextAnalyzerException("Unable to analyze element '$image' using a remote analysis service",
                            e)
                }

            }
        } finally {
            closeConnection(response)
        }
        throw TextAnalyzerException("Unable to analyze element '$image' using a remote analysis service")
    }

    companion object {

        /** The logger  */
        private val logger = LoggerFactory.getLogger(TextAnalysisRemoteImpl::class.java)
    }

}
